server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/cs/factory/DeployedCacheStoreFactory.java

```diff
@@ -18,49 +21,68 @@
  */
 public class DeployedCacheStoreFactory implements CacheStoreFactory {
 
-   private Map<String, DeployedCacheStoreMetadata> deployedCacheStores = Collections.synchronizedMap(new HashMap<String, DeployedCacheStoreMetadata>());
+   private static final int TIMEOUT_SECONDS = 60;
+
+   private ConcurrentHashMap<String, CompletableFuture<DeployedCacheStoreMetadata>> deployedCacheStores = new ConcurrentHashMap<>();
 
    @Override
    public <T> T createInstance(StoreConfiguration cfg) {
-      if(cfg instanceof DeployedStoreConfiguration) {
+      if (cfg instanceof DeployedStoreConfiguration) {
          DeployedStoreConfiguration deployedConfiguration = (DeployedStoreConfiguration) cfg;
-         DeployedCacheStoreMetadata deployedCacheStoreMetadata = deployedCacheStores.get(deployedConfiguration.getCustomStoreClassName());
-         return (T) deployedCacheStoreMetadata.getLoaderWriterRawInstance();
+         DeployedCacheStoreMetadata deployedCacheStoreMetadata = null;
+         try {
+            deployedCacheStoreMetadata = getPromise(deployedConfiguration.getCustomStoreClassName()).get(TIMEOUT_SECONDS, TimeUnit.SECONDS);
+            return (T) deployedCacheStoreMetadata.getLoaderWriterRawInstance();
+         } catch (InterruptedException e) {
+            InfinispanLogger.ROOT_LOGGER.error("Could not get Custom Cache Store metadata (" + deployedConfiguration.getCustomStoreClassName() + ") because of an InterruptionException.", e);
+         } catch (ExecutionException e) {
+            InfinispanLogger.ROOT_LOGGER.error("Could not get Custom Cache Store metadata (" + deployedConfiguration.getCustomStoreClassName() + ") because of an ExecutionException.", e);
```

Comment:
It depends how we solve the above question...

If we decided to leave it as is, I would prefer to leave it wrapped inside `ExecutionException`. If we decide to throw a `RuntimeException`, than I would prefer to rewrap it to `RuntimeException` and rethrow it. Let's wait and see if Tristan has some ideas...